Cgen erorr when assign `fn (s string) {}` to field `bar ?fn (string)`

### Describe the bug

Code:
```v
module main

struct Foo {
mut:
	bar ?fn (string)
}

mut foo := Foo{}
foo.bar = fn (s string) {  }
```

### Expected Behavior

No error

### Current Behavior

Error:
```

==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/main.11010486774995230345.tmp.c:12294:10: error: assigning to '_option_anon_fn_string' (aka 'struct _option_anon_fn_string') from incompatible type 'void (string)' (aka 'void (struct string)')
        foo.bar =       anon_fn_a281bdd8f981808c_string_93;
                ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

Run code above.

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 e8ca2e6

### Environment details (OS name and version, etc.)

```
OS: macos, macOS, 13.0.1, 22A400
Processor: 10 cpus, 64bit, little endian, Apple M1 Pro
CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)

getwd: /Users/petrmakhnev/v
vmodules: /Users/petrmakhnev/.vmodules
vroot: /Users/petrmakhnev/v
vexe: /Users/petrmakhnev/v/v
vexe mtime: 2023-02-04 16:00:37
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.3 e8ca2e6

Git version: git version 2.37.1 (Apple Git-137.1)
Git vroot status: weekly.2023.05-5-ge8ca2e62
.git/config present: true
```

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -162,36 +162,38 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 			}
 			// TODO: no buffer fiddling
 			ast.AnonFn {
-				if blank_assign {
-					g.write('{')
-				}
-				// if it's a decl assign (`:=`) or a blank assignment `_ =`/`_ :=` then generate `void (*ident) (args) =`
-				if (is_decl || blank_assign) && left is ast.Ident {
-					sig := g.fn_var_signature(val.decl.return_type, val.decl.params.map(it.typ),
-						ident.name)
-					g.write(sig + ' = ')
-				} else {
-					g.is_assign_lhs = true
-					g.assign_op = node.op
-					g.expr(left)
-					g.is_assign_lhs = false
-					g.is_arraymap_set = false
-					if mut left is ast.IndexExpr {
-						sym := g.table.sym(left.left_type)
-						if sym.kind in [.map, .array] {
-							g.expr(val)
-							g.writeln('});')
-							continue
-						}
+				if !(var_type.has_flag(.option) || var_type.has_flag(.result)) {
+					if blank_assign {
+						g.write('{')
 					}
-					g.write(' = ')
+					// if it's a decl assign (`:=`) or a blank assignment `_ =`/`_ :=` then generate `void (*ident) (args) =`
+					if (is_decl || blank_assign) && left is ast.Ident {
+						sig := g.fn_var_signature(val.decl.return_type, val.decl.params.map(it.typ),
+							ident.name)
+						g.write(sig + ' = ')
+					} else {
+						g.is_assign_lhs = true
+						g.assign_op = node.op
+						g.expr(left)
+						g.is_assign_lhs = false
+						g.is_arraymap_set = false
+						if mut left is ast.IndexExpr {
+							sym := g.table.sym(left.left_type)
+							if sym.kind in [.map, .array] {
+								g.expr(val)
+								g.writeln('});')
+								continue
+							}
+						}
+						g.write(' = ')
+					}
+					g.expr(val)
+					g.writeln(';')
+					if blank_assign {
+						g.write('}')
+					}
+					continue
 				}
-				g.expr(val)
-				g.writeln(';')
-				if blank_assign {
-					g.write('}')
-				}
-				continue
 			}
 			else {}
 		}

--- a/vlib/v/tests/assign_option_of_struct_fntype_field_test.v
+++ b/vlib/v/tests/assign_option_of_struct_fntype_field_test.v
@@ -1,0 +1,16 @@
+struct Foo {
+mut:
+	bar ?fn (string) string
+}
+
+fn test_assign_option_of_struct_fntype_field() {
+	mut foo := Foo{}
+	foo.bar = fn (s string) string {
+		return s
+	}
+	if ff := foo.bar {
+		ret := ff('hello')
+		println(ret)
+		assert ret == 'hello'
+	}
+}

